pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/AvoidInstantiatingObjectsInLoopsRule.java

```diff
@@ -14,24 +14,77 @@
 import net.sourceforge.pmd.lang.java.ast.ASTWhileStatement;
 
 public class AvoidInstantiatingObjectsInLoopsRule extends AbstractOptimizationRule {
-
+    /**
+     * This method is used to check whether user instantiates variables
+     * which are not assigned in loops.
+     * @param node This is the expression of part of java code to be checked.
+     * @param data This is the data to return.
+     * @return Object This returns the data passed in. If violation happens, violation is added to data.
+     */
     @Override
     public Object visit(ASTAllocationExpression node, Object data) {
+        //CS304 Issue link: https://github.com/pmd/pmd/issues/2207
         if (insideLoop(node) && fourthParentNotThrow(node) && fourthParentNotReturn(node)) {
-            addViolation(data, node);
+            if (thirdParentNotASTExpression(node) && fourthParentNotASTStatementExpression(node)) {
+                addViolation(data, node);
+            }
         }
         return data;
     }
 
-    private boolean fourthParentNotThrow(ASTAllocationExpression node) {
+    /**
+     * This method is used to check whether the instantiated variable is assigned or not.
+     * @param node This is the expression of part of java code to be checked.
+     * @return boolean This returns Whether the third parent of node is an ASTExpression.
+     */
+    public boolean thirdParentNotASTExpression(ASTAllocationExpression node) {
+        //CS304 Issue link: https://github.com/pmd/pmd/issues/2207
+        if (node.getParent().getClass().toString().equals(
+                "class net.sourceforge.pmd.lang.java.ast.ASTPrimaryPrefix")
+                && node.getParent().getParent().getClass().toString().equals(
+                    "class net.sourceforge.pmd.lang.java.ast.ASTPrimaryExpression")) {
+            return !node.getParent().getParent().getParent().getClass().toString().equals(
+                    "class net.sourceforge.pmd.lang.java.ast.ASTExpression");
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * This method is used to check whether the instantiated variable is assigned or not.
+     * @param node This is the expression of part of java code to be checked.
+     * @return boolean This returns Whether the fourth parent of node is an ASTStatementExpression.
+     */
+    public boolean fourthParentNotASTStatementExpression(ASTAllocationExpression node) {
+        //CS304 Issue link: https://github.com/pmd/pmd/issues/2207
+        return !node.getParent().getParent().getParent().getClass().toString().equals(
+                "class net.sourceforge.pmd.lang.java.ast.ASTStateExpression");
```

Comment:
There is no class ASTStateExpression - so this equals always return "true"...
Did you mean "ASTStatementExpression"?
That's by the way one example, why using toString is bad: It didn't show you this error during compile time...